File: binary_search/search_in_rotated_sorted_array/js/search_in_rotated_sorted_array.js
const nums = [4, 5, 6, 7, 0, 1, 2];
const target = 0;

const search = (nums, target) => {
  let l = 0;
  let r = nums.length-1;

  while(l <= r) {
    let m = (l + r) >> 1;
    if(nums[m] === target) return m;
    if(nums[l] <= nums[m]) {
      if(nums[l] <= target && target <= nums[m]) {
        r = m-1;
      } else {
        l = m+1;
      }
    } else {
      if(nums[m] <= target && target <= nums[r]) {
        l = m+1;
      } else {
        r = m-1;
      }
    }
  }
  return -1;
}

console.log(search(nums, target));

File: binary_search/search_in_rotated_sorted_array/cpp/search_in_rotated_sorted_array.cpp
#include<iostream>
#include<vector>

using std::vector;

// Time O(log(n)) 
// Space O(1)

class Solution {
public:
    static int search(const vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size()-1;

        if(nums.size() == 1 && nums[0] == target) { return 0; }

        while(l <= r) {
            int m = (l + r) >> 1;
            if(nums[m] == target) return m;
            if(nums[l] <= nums[m]) {
                if(nums[l] <= target && target <= nums[m]) {
                    r = m-1;
                } else {
                    l = m+1;
                }
            } else {
                if(nums[m] <= target && target <= nums[r]) {
                    l = m+1;
                } else {
                    r = m-1;
                }
            }
        }

        return -1;
    }
};

int main() {
    std::vector<int> nums = { 4, 5, 6, 7, 0, 1, 2 };
    int target = 0;

    int answer = Solution::search(nums, target);

    std::cout << answer << '\n';

    return 0;
}
